src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -150,6 +150,11 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
+            // We can cancel at entry, but once the pipe operations are scheduled we rely on both operations running to
+            // avoid deadlocks (the exception handler in 'writerTask' ensures progress is made in 'readerTask').
+            cancellationToken.ThrowIfCancellationRequested();
+            var mustNotCancelToken = CancellationToken.None;
```

Comment:
i don't have an issue with your fix, but this seems super hard to get right with pipes.  Is this just what everyone needs to deal with?  Is there no better way to have bidi pipes with cancellation?

Reply:
The complexity is encapsulated, so for me there are two lessons:

1. Keep it encapsulated.
2. Use caution before moving forward with a change proposal to use System.IO.Pipelines in new places. It hard.

Reply:
sgtm.  if you get a good grasp on how to write things correctly, a message/gist in IDE channel might be helpful :)

Reply:
(in the future of course) :)

Reply:
The general rule is that pipe readers and writers must be completed. Readers and writers passed to JSON-RPC as remote call arguments are completed by JSON-RPC. The problem is to figure out all the possible code paths in presence of exceptions and cancellations.